internal/venice-common/src/main/java/com/linkedin/venice/hooks/StoreLifecycleHooks.java

```diff
@@ -0,0 +1,311 @@
+ * is passed into all hooks. Note that there is just a single bag of properties per store, therefore the configs of all
+ * registered hooks must coexist within that bag. As such, namespacing is strongly recommended, so that different hooks
+ * don't conflict with each other.<br>
+ * <br>
+ * <b>Cardinality</b><br>
+ * <br>
+ * Each hook is annotated with a cardinality, providing a rough indication of how often the hook can be expected to be
+ * invoked.<br>
+ * <br>
+ * <b>Thread safety</b><br>
+ * <br>
+ * Hook functions can be invoked concurrently and so hook implementations are expected to be thread-safe.<br>
+ * <br>
+ * However, the hooks framework guarantees to call hooks one at a time for a given store, so even if multiple hooks for
+ * a given store simultaneously become eligible to be called, their invocation will be serialized.<br>
+ * @see StoreLifecycleEventOutcome
+ * @see StoreVersionLifecycleEventOutcome
+ */
+@Threadsafe
+public abstract class StoreLifecycleHooks {
+  private final VeniceProperties defaultConfigs;
+
+  public StoreLifecycleHooks(VeniceProperties defaultConfigs) {
+    this.defaultConfigs = defaultConfigs;
+  }
+
+  protected VeniceProperties getDefaultConfigs() {
+    return this.defaultConfigs;
+  }
+
+  /**
+   * Invoked prior to updating a store's config with a given hook config. If the hook implementation recognizes the
+   * config key and the associated config value is valid (e.g. can be parsed or cast to the correct type), then it
+   * should return {@link StoreLifecycleEventOutcome#PROCEED}, otherwise it should return
+   * {@link StoreLifecycleEventOutcome#ABORT}.<br>
+   * <br>
+   * Note that there is a single hook config bag shared by all registered hooks, and so config keys should be namespaced
+   * in order to avoid collisions (unless a given config key is intended to control similar behavior in many hooks).<br>
+   * <br>
+   * The hooks framework will reject the store config update if all registered hooks return the ABORT signal, and will
+   * accept it if at least one registered hook returns the PROCEED signal.<br>
+   * <br>
+   * Cardinality: once per config key/value pair included in a store update command.
+   */
+  public StoreLifecycleEventOutcome validateHookConfig(
+      String clusterName,
+      String storeName,
+      String configKey,
+      String configValue) {
+    return StoreLifecycleEventOutcome.PROCEED;
+  }
+
+  /**
+   * Invoked prior to starting a new job. The hook has the option of aborting the job.<br>
+   * <br>
+   * N.B.: this hook returns a {@link StoreLifecycleEventOutcome}, and not a {@link StoreVersionLifecycleEventOutcome},
+   * because the new store-version is not created yet at the start of the push job.<br>
+   * <br>
+   * Cardinality: once per push job.
+   */
+  public StoreLifecycleEventOutcome preStartOfPushJob(
+      String clusterName,
+      String storeName,
+      VeniceProperties jobProperties,
```

Comment:
If we decide to execute the hooks in parent controller, do we have a way to grab the push job properties?

Reply:
I thought it might be useful for the hook to know the properties, though I'm not sure that we have a concrete use case that must have that as of today.

If we want to keep this parameter here, and execute the hooks in the controller, then we may need to pass the job props to the `/request_topic` endpoint. Alternatively, the job props are in the `PushJobDetails` system store and we may be able to query them out of there (in which case, maybe we should wrap it inside a `Lazy` so that it's only queried if the hook needs it).

Reply:
Since there is no known use case requires this, shall we remove it?

Reply:
Done!